File: dart/lib/src/utils/deserialize_record_content.dart
import 'dart:convert';
import 'dart:typed_data';

import '../constants/records.dart';
import '../errors/sns_errors.dart';
import 'base58_utils.dart';

/// Parameters for deserializing record content
class DeserializeRecordContentParams {
  const DeserializeRecordContentParams({
    required this.content,
    required this.record,
  });

  /// The content to deserialize
  final Uint8List content;

  /// The type of record
  final Record record;
}

/// Deserialize record content based on the record type
///
/// This function mirrors js-kit/src/utils/deserializers/deserializeRecordContent.ts
///
/// [content] - The content to deserialize
/// [record] - The type of record
///
/// Returns the deserialized content as a string
String deserializeRecordContent({
  required Uint8List content,
  required Record record,
}) {
  final isUtf8Encoded = utf8EncodedRecords.contains(record);

  if (isUtf8Encoded) {
    final decoded = utf8.decode(content);
    if (record == Record.cname || record == Record.txt) {
      // In a full implementation, this would use punycode decoding
      // For now, return the decoded UTF-8 string
      return decoded;
    }
    return decoded;
  } else if (record == Record.sol) {
    return _base58Encode(content);
  } else if (evmRecords.contains(record)) {
    return '0x${_uint8ArrayToHex(content)}';
  } else if (record == Record.injective) {
    // In a full implementation, this would use bech32 encoding
    // For now, return a placeholder
    return 'inj${_uint8ArrayToHex(content)}';
  } else if (record == Record.a || record == Record.aaaa) {
    // IP address deserialization
    if (record == Record.a && content.length == 4) {
      return '${content[0]}.${content[1]}.${content[2]}.${content[3]}';
    } else if (record == Record.aaaa && content.length == 16) {
      final groups = <String>[];
      for (var i = 0; i < 16; i += 2) {
        final group = (content[i] << 8) | content[i + 1];
        groups.add(group.toRadixString(16));
      }
      return groups.join(':');
    } else {
      throw InvalidRecordDataError('Invalid IP address record length');
    }
  } else {
    throw InvalidRecordDataError('The record content is malformed');
  }
}

/// Convert uint8 array to hex string
String _uint8ArrayToHex(Uint8List bytes) =>
    bytes.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join();

/// Base58 encode helper
String _base58Encode(Uint8List bytes) => Base58Utils.encode(bytes);

File: js/src/record/deserializeRecordContent.ts
import ipaddr from "ipaddr.js";
import punycode from "punycode/punycode.js";

import { bech32 } from "@scure/base";
import { PublicKey } from "@solana/web3.js";

import { InvalidRecordDataError } from "../error";
import { Record } from "../types/record";
import { EVM_RECORDS, UTF8_ENCODED } from "./const";

import type { Buffer } from "buffer";

/**
 * Deserializes record content according to SNS-IP 1.
 *
 * @param content Serialized record content
 * @param record Record type
 * @returns Deserialized record content.
 */
export const deserializeRecordContent = (
  content: Buffer,
  record: Record,
): string => {
  const utf8Encoded = UTF8_ENCODED.has(record);

  if (utf8Encoded) {
    const decoded = content.toString("utf-8");
    if (record === Record.CNAME || record === Record.TXT) {
      return punycode.decode(decoded);
    }
    return decoded;
  } else if (record === Record.SOL) {
    return new PublicKey(content).toBase58();
  } else if (EVM_RECORDS.has(record)) {
    return "0x" + content.toString("hex");
  } else if (record === Record.Injective) {
    return bech32.encode("inj", bech32.toWords(content));
  } else if (record === Record.A || record === Record.AAAA) {
    return ipaddr.fromByteArray([...content]).toString();
  } else {
    throw new InvalidRecordDataError("The record content is malformed");
  }
};
